Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2ClientRequestHandler.swift

```diff
@@ -0,0 +1,330 @@
+    typealias InboundIn = HTTPClientResponsePart
+
+    private let eventLoop: EventLoop
+
+    private var state: HTTPRequestStateMachine = .init(isChannelWritable: false) {
+        didSet {
+            self.eventLoop.assertInEventLoop()
+        }
+    }
+
+    /// while we are in a channel pipeline, this context can be used.
+    private var channelContext: ChannelHandlerContext?
+
+    private var request: HTTPExecutableRequest? {
+        didSet {
+            if let newRequest = self.request, let idleReadTimeout = newRequest.idleReadTimeout {
+                self.idleReadTimeoutStateMachine = .init(timeAmount: idleReadTimeout)
+            } else {
+                self.idleReadTimeoutStateMachine = nil
+            }
+        }
+    }
+
+    private var idleReadTimeoutStateMachine: IdleReadStateMachine?
+    private var idleReadTimeoutTimer: Scheduled<Void>?
+
+    init(eventLoop: EventLoop) {
+        self.eventLoop = eventLoop
+    }
+
+    func handlerAdded(context: ChannelHandlerContext) {
+        self.channelContext = context
+
+        let isWritable = context.channel.isActive && context.channel.isWritable
+        let action = self.state.writabilityChanged(writable: isWritable)
+        self.run(action, context: context)
+    }
+
+    func handlerRemoved(context: ChannelHandlerContext) {
+        self.channelContext = nil
+    }
+
+    // MARK: Channel Inbound Handler
+
+    func channelActive(context: ChannelHandlerContext) {
+        let action = self.state.writabilityChanged(writable: context.channel.isWritable)
+        self.run(action, context: context)
+    }
+
+    func channelInactive(context: ChannelHandlerContext) {
+        let action = self.state.channelInactive()
+        self.run(action, context: context)
+    }
+
+    func channelWritabilityChanged(context: ChannelHandlerContext) {
+        let action = self.state.writabilityChanged(writable: context.channel.isWritable)
+        self.run(action, context: context)
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let httpPart = self.unwrapInboundIn(data)
+
+        if let timeoutAction = self.idleReadTimeoutStateMachine?.channelRead(httpPart) {
+            self.runTimeoutAction(timeoutAction, context: context)
+        }
+
+        let action = self.state.channelRead(httpPart)
+        self.run(action, context: context)
+    }
+
+    func channelReadComplete(context: ChannelHandlerContext) {
+        let action = self.state.channelReadComplete()
+        self.run(action, context: context)
+    }
+
+    func errorCaught(context: ChannelHandlerContext, error: Error) {
+        let action = self.state.errorHappened(error)
+        self.run(action, context: context)
+    }
+
+    // MARK: Channel Outbound Handler
+
+    func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        let request = self.unwrapOutboundIn(data)
+        self.request = request
```

Comment:
Should we be validating that `self.request` is `nil` here?

Reply:
Added a comment, that this check is done by the statemachine.